federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package federatedtypes
+
+import (
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/watch"
+	federationclientset "k8s.io/kubernetes/federation/client/clientset_generated/federation_clientset"
+	"k8s.io/kubernetes/federation/pkg/federation-controller/util"
+	"k8s.io/kubernetes/pkg/api"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	autoscalingv1 "k8s.io/kubernetes/pkg/apis/autoscaling/v1"
+	kubeclientset "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+)
+
+const (
+	HpaKind              = "horizontalpodautoscaler"
+	HpaControllerName    = "horizontalpodautoscalers"
+	scaleForbiddenWindow = 5 * time.Minute
+)
+
+func init() {
+	RegisterFederatedType(HpaKind, HpaControllerName, []schema.GroupVersionResource{apiv1.SchemeGroupVersion.WithResource(HpaControllerName)}, NewHpaAdapter)
+}
+
+type HpaAdapter struct {
+	client federationclientset.Interface
+}
+
+func NewHpaAdapter(client federationclientset.Interface) FederatedTypeAdapter {
+	return &HpaAdapter{client: client}
+}
+
+func (a *HpaAdapter) Kind() string {
+	return HpaKind
+}
+
+func (a *HpaAdapter) ObjectType() pkgruntime.Object {
+	return &autoscalingv1.HorizontalPodAutoscaler{}
+}
+
+func (a *HpaAdapter) IsExpectedType(obj interface{}) bool {
+	_, ok := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	return ok
+}
+
+func (a *HpaAdapter) Copy(obj pkgruntime.Object) pkgruntime.Object {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	return &autoscalingv1.HorizontalPodAutoscaler{
+		ObjectMeta: util.DeepCopyRelevantObjectMeta(hpa.ObjectMeta),
+		Spec:       *(util.DeepCopyApiTypeOrPanic(&hpa.Spec).(*autoscalingv1.HorizontalPodAutoscalerSpec)),
+	}
+}
+
+func (a *HpaAdapter) Equivalent(obj1, obj2 pkgruntime.Object) bool {
+	return util.ObjectMetaAndSpecEquivalent(obj1, obj2)
+}
+
+func (a *HpaAdapter) NamespacedName(obj pkgruntime.Object) types.NamespacedName {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
+	return types.NamespacedName{Namespace: hpa.Namespace, Name: hpa.Name}
+}
+
+func (a *HpaAdapter) ObjectMeta(obj pkgruntime.Object) *metav1.ObjectMeta {
+	return &obj.(*autoscalingv1.HorizontalPodAutoscaler).ObjectMeta
+}
+
+func (a *HpaAdapter) FedCreate(obj pkgruntime.Object) (pkgruntime.Object, error) {
+	hpa := obj.(*autoscalingv1.HorizontalPodAutoscaler)
```

Comment:
Catching this failure here and at all other places which use runtime.object and then cast to a specific type might be considered as being extremely safe programming local to the function. But if this happens then there probably is much bigger a problem which needs to be looked into and the tests I guess can catch that, for sure. Also, this is same as done for all other controllers using sync controller.